src/recluster.c

```diff
@@ -0,0 +1,1140 @@
+
+	/* Reset rd_toastoid just to be tidy --- it shouldn't be looked at again */
+	NewHeap->rd_toastoid = InvalidOid;
+
+	num_pages = RelationGetNumberOfBlocks(NewHeap);
+
+	/* Log what we did */
+	ereport(elevel,
+			(errmsg("\"%s\": found %.0f removable, %.0f nonremovable row versions in %u pages",
+					RelationGetRelationName(OldHeap),
+					tups_vacuumed, num_tuples,
+					RelationGetNumberOfBlocks(OldHeap)),
+			 errdetail("%.0f dead row versions cannot be removed yet.\n"
+					   "%s.",
+					   tups_recently_dead,
+					   pg_rusage_show(&ru0))));
+
+	/* Clean up */
+	pfree(values);
+	pfree(isnull);
+
+	if (OldIndex != NULL)
+		index_close(OldIndex, NoLock);
+	heap_close(OldHeap, NoLock);
+	heap_close(NewHeap, NoLock);
+
+	/* Update pg_class to reflect the correct values of pages and tuples. */
+	relRelation = heap_open(RelationRelationId, RowExclusiveLock);
+
+	reltup = SearchSysCacheCopy1(RELOID, ObjectIdGetDatum(OIDNewHeap));
+	if (!HeapTupleIsValid(reltup))
+		elog(ERROR, "cache lookup failed for relation %u", OIDNewHeap);
+	relform = (Form_pg_class) GETSTRUCT(reltup);
+
+	relform->relpages = num_pages;
+	relform->reltuples = num_tuples;
+
+	/* Don't update the stats for pg_class.  See swap_relation_files. */
+	if (OIDOldHeap != RelationRelationId)
+		CatalogTupleUpdate(relRelation, &reltup->t_self, reltup);
+	else
+		CacheInvalidateRelcacheByTuple(reltup);
+
+	/* Clean up. */
+	heap_freetuple(reltup);
+	heap_close(relRelation, RowExclusiveLock);
+
+	/* Make the update visible */
+	CommandCounterIncrement();
+}
+
+
+
+/*
+ * Remove the transient table that was built by make_new_heap, and finish
+ * cleaning up (including rebuilding all indexes on the old heap).
+ *
+ * NB: new_index_oids must be in the same order as RelationGetIndexList
+ *
+ */
+static void
+finish_heap_swaps(Oid OIDOldHeap, Oid OIDNewHeap,
+				  List *old_index_oids,
+				  List *new_index_oids,
+				  bool swap_toast_by_content,
+				  bool is_internal,
+				  TransactionId frozenXid,
+				  MultiXactId cutoffMulti)
+{
+	ObjectAddress object;
+	Relation	oldHeapRel;
+	ListCell   *old_index_cell;
+	ListCell   *new_index_cell;
+
+
+	/*
+	 * There's a risk of deadlock if some other process is also trying to
+	 * upgrade their lock in the same manner as us, at this time. Since our
+	 * transaction has performed a large amount of work, and only needs to be
+	 * run once per chunk, we do not want to abort it due to this deadlock. To
+	 * prevent abort we set our `deadlock_timeout` to a large value in the
+	 * expectation that the other process will timeout and abort first.
+	 * Currently we set `deadlock_timeout` to 1 hour, as this should be longer
+	 * than any other normal process, while still allowing the system to make
+	 * progress in the event of a real deadlock. As this is the last lock we
+	 * grab, and the setting is local to our transaction we do not bother
+	 * changing the guc back.
+	 */
+	int			config_change = set_config_option("deadlock_timeout",
```

Comment:
Should we undo this change at any point?

Reply:
We don't grab any locks for the remainder of the transaction, and change is transaction local, and setting it to the previous value would involve round tripping it through a string.